In [5]:
import torch
from torch.utils.data import Dataset, DataLoader
import torch.nn as nn
import torch.nn.functional as F
from torch.optim import Adam

import numpy as np
from sklearn.metrics import confusion_matrix

import os
import pandas as pd
import time
import warnings
from tqdm.notebook import tqdm

import librosa
import matplotlib.pyplot as plt

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
#device = torch.device("cpu")
print(f"Using device: {device}")

print(torch.cuda.device_count())
print(torch.cuda.get_device_name())

warnings.filterwarnings('ignore')

Using device: cuda
2
Quadro GP100


In [6]:
class FeatureDataset(Dataset):
    def __init__(self, csv_file, transform=None, target_transform=None):
        self.df = pd.read_csv(csv_file)
        self.transform = transform
        self.target_transform = target_transform
        self.label_map = {
            'Drone': 0,
            'No Drone': 1
        }

    def __len__(self):
        return len(self.df)

    def __getitem__(self, idx):
        if torch.is_tensor(idx):
            idx = idx.tolist()
        features = torch.FloatTensor(self.df.iloc[idx, 2:])
        label = self.df.iloc[idx, 1]
        
        if self.target_transform:
            label = self.target_transform(label)
        
        return torch.tensor(features), torch.tensor(label)
    

# Usage example
csv_file = "/home/distasiom/Documents/Summer2024/data/complete_dataset_features.csv"

# Define any transformations if needed
transform = None
target_transform = None

dataset = FeatureDataset(csv_file, transform=transform, target_transform=target_transform)

In [7]:
train_size = int(.8 * len(dataset))
test_size = int(.75 * (len(dataset) - train_size))
val_size = len(dataset) - train_size - test_size

train_dataset, test_dataset, val_dataset = torch.utils.data.random_split(dataset, [train_size, test_size, val_size])
print(train_dataset)
print(train_dataset[0])

train_loader = DataLoader(
    train_dataset,
    batch_size=64,
    shuffle=True
    )

test_loader = DataLoader(
    test_dataset,
    batch_size=64,
    shuffle=False
)

val_loader = DataLoader(
    val_dataset,
    batch_size=64,
    shuffle=False
)

(tensor([-141.4258, -106.4519, -123.4676,  ...,    0.2202,    0.2202,
           0.2202]), tensor(1))


In [ ]:
for feature, label in train_loader:
    print(label)

In [ ]:
print(train_dataset[7][0].shape)


In [ ]:
# Run 1 small model (36,606,465 parameters)
class LNN(nn.Module):
    def __init__(self):
        super().__init__()
        self.seq = nn.Sequential(
            nn.LazyLinear(4096),
            nn.LazyBatchNorm1d(),
            nn.ReLU(),
            nn.LazyLinear(4096),
            nn.LazyBatchNorm1d(),
            nn.ReLU(),
            nn.LazyLinear(1024),
            nn.LazyBatchNorm1d(),
            nn.ReLU(),
            nn.LazyLinear(1024),
            nn.LazyBatchNorm1d(),
            nn.ReLU(),
            nn.LazyLinear(512),
            nn.LazyBatchNorm1d(),
            nn.ReLU(),
            nn.LazyLinear(256),
            nn.LazyBatchNorm1d(),
            nn.ReLU(),
            nn.LazyLinear(1)
        )

    def forward(self, x):
        logits = self.seq(x).squeeze(dim=1)
        return logits


In [8]:
# Run 2 larger model (47,104,513 parameters)
class LNN(nn.Module):
    def __init__(self):
        super().__init__()
        self.seq = nn.Sequential(
            nn.LazyLinear(4096),
            nn.LazyBatchNorm1d(),
            nn.ReLU(),
            nn.LazyLinear(4096),
            nn.LazyBatchNorm1d(),
            nn.ReLU(),
            nn.LazyLinear(2048),
            nn.LazyBatchNorm1d(),
            nn.ReLU(),
            nn.LazyLinear(2048),
            nn.LazyBatchNorm1d(),
            nn.ReLU(),
            nn.LazyLinear(1024),
            nn.LazyBatchNorm1d(),
            nn.ReLU(),
            nn.LazyLinear(1024),
            nn.LazyBatchNorm1d(),
            nn.ReLU(),
            nn.LazyLinear(512),
            nn.LazyBatchNorm1d(),
            nn.ReLU(),
            nn.LazyLinear(256),
            nn.LazyBatchNorm1d(),
            nn.ReLU(),
            nn.LazyLinear(1)
        )

    def forward(self, x):
        logits = self.seq(x).squeeze(dim=1)
        return logits


In [9]:
def split_seconds(seconds):
    minutes = seconds // 60
    hours = minutes // 60
    days = hours // 24
    return seconds % 60, minutes % 60, hours % 24, days


In [10]:
def main():
    # Define model
    model = LNN()

    # Cuda setup
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    # device = torch.device("cpu")
    model = model.to(device)
   
    print(f"Using device: {device}")
    # Optimizer setup
    optimizer = Adam(model.parameters(), lr=1e-4)
   
    # Loss function
    loss_fn = nn.BCEWithLogitsLoss(reduction="mean")
   
    # Number of epochs
    num_epochs = 64
   
    # Train or load model?
    print("Training model....")
    start = time.time()
    for epoch in tqdm(range(num_epochs), desc="Epochs"):
        
        model.train()
        for batch_idx, (features, labels) in enumerate(tqdm(train_loader, desc=f"Epoch {epoch+1}/{num_epochs}", leave=False)):
            features, labels = features.to(device), labels.to(device)
            optimizer.zero_grad()
           
            # CNN forward pass
            logits = model(features)
            loss = loss_fn(logits, labels.to(float))
            loss.backward()
            optimizer.step()

        model.eval()
        num_test = 0
        num_correct = 0
        total_loss = 0
        with torch.no_grad():
            for features, labels in tqdm(test_loader, desc="Testing"):
                features, labels = features.to(device), labels.to(device)
                logits = model(features)
                loss = loss_fn(logits, labels.to(float))
                total_loss += loss.item()
                preds = torch.round(torch.sigmoid(logits))
                num_test += labels.size(0)
                num_correct += preds.eq(labels).sum().item()
    
        tqdm.write(f"Epoch {epoch+1}/{num_epochs}, Test accuracy: {num_correct / num_test * 100:.2f}, Total loss: {total_loss}")

#         torch.save(model.state_dict(), f"/home/distasiom/Documents/Summer2024/models/dnd_mfcc_classifier/epoch_{str(epoch)}.pt")
   
    end = time.time()
    seconds, minutes, hours, days = split_seconds(end - start)
    print(f"Training Runtime: {int(days)}d {int(hours)}h {int(minutes)}m {seconds:.2f}s")
   
    # Evaluate model on validation data
    model.eval()
    print("Evaluating model....")
    start = time.time()
    total_loss = 0
    num_test = 0
    num_correct = 0
   
    with torch.no_grad():
        for features, labels in tqdm(val_loader, desc="Validating"):
            features, labels = features.to(device), labels.to(device)
            logits = model(features)
            loss = loss_fn(logits, labels.to(float))
            total_loss += loss.item()
            preds = torch.round(torch.sigmoid(logits))
            num_test += labels.size(0)
            num_correct += preds.eq(labels).sum().item()
   
    print(f"Test accuracy: {num_correct / num_test * 100:.2f}, Total loss: {total_loss}")
    end = time.time()
    seconds, minutes, hours, days = split_seconds(end - start)
    print(f"Testing Runtime: {int(days)}d {int(hours)}h {int(minutes)}m {seconds:.2f}s")

In [ ]:
main()

In [11]:
# Print number of parameters
model = LNN()
state = torch.load("/home/distasiom/Documents/Summer2024/models/dnd_mfcc_classifier_large/epoch_0.pt")
model.load_state_dict(state)
total_params = sum([torch.numel(p) for p in model.parameters()])
print(f"Total parameters: {total_params}")

Total parameters: 47104513


In [12]:
# Define model
model = LNN()
# Cuda setup
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
# device = torch.device("cpu")
model = model.to(device)

print(f"Using device: {device}")

directory = "/home/distasiom/Documents/Summer2024/models/dnd_mfcc_classifier_large"
acc_epoch = None
max_acc = 0
loss_epoch = None
min_loss = float('inf')

for file in os.listdir(directory):
    # Optimizer setup
    optimizer = Adam(model.parameters(), lr=1e-3)

    # Loss function
    loss_fn = nn.BCEWithLogitsLoss(reduction="mean")

    # Load model
    filepath = os.path.join(directory, file)
    state = torch.load(filepath)
    model.load_state_dict(state)

    # Evaluate model on validation data
    model.eval()
    print(f"Evaluating model {file}")
    total_loss = 0
    num_test = 0
    num_correct = 0

    with torch.no_grad():
        for features, labels in tqdm(val_loader, desc="Validating"):
            features, labels = features.to(device), labels.to(device)
            logits = model(features)
            loss = loss_fn(logits, labels.to(float))
            total_loss += loss.item()
            preds = torch.round(torch.sigmoid(logits))
            num_test += labels.size(0)
            num_correct += preds.eq(labels).sum().item()

    accuracy = num_correct / num_test * 100
    if accuracy > max_acc:
        max_acc = accuracy
        acc_epoch = file
    if total_loss < min_loss:
        min_loss = total_loss
        loss_epoch = file

    print(f"Test accuracy: {accuracy:.2f}, Total loss: {total_loss}")

print(f"Best accuracy: {max_acc} at {acc_epoch}")
print(f"Best loss: {min_loss} at {loss_epoch}")

Using device: cuda
Evaluating model epoch_0.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 89.52, Total loss: 12.015218783603684
Evaluating model epoch_1.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 90.21, Total loss: 10.812226473866227
Evaluating model epoch_2.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 90.38, Total loss: 10.816444699581364
Evaluating model epoch_3.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 90.24, Total loss: 11.19993125260458
Evaluating model epoch_4.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 90.24, Total loss: 10.670400701421098
Evaluating model epoch_5.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 91.05, Total loss: 9.736568362088398
Evaluating model epoch_6.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 91.25, Total loss: 9.78154882766374
Evaluating model epoch_7.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 91.57, Total loss: 9.383338392774462
Evaluating model epoch_8.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 91.02, Total loss: 8.999296605018433
Evaluating model epoch_9.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 92.00, Total loss: 8.763106389009
Evaluating model epoch_10.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 92.52, Total loss: 8.513359007802254
Evaluating model epoch_11.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 91.97, Total loss: 8.647671673134314
Evaluating model epoch_12.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 92.20, Total loss: 8.369289457061297
Evaluating model epoch_13.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 92.61, Total loss: 7.914787584041533
Evaluating model epoch_14.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 93.07, Total loss: 7.8560673528891884
Evaluating model epoch_15.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 93.42, Total loss: 7.197807668273266
Evaluating model epoch_16.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 93.10, Total loss: 7.551542014369516
Evaluating model epoch_17.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 93.24, Total loss: 7.812840938276682
Evaluating model epoch_18.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 93.44, Total loss: 7.513110080184688
Evaluating model epoch_19.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 92.90, Total loss: 7.585653714101114
Evaluating model epoch_20.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.08, Total loss: 7.036919210581822
Evaluating model epoch_21.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 93.53, Total loss: 7.221238494238017
Evaluating model epoch_22.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.08, Total loss: 7.184732316132925
Evaluating model epoch_23.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 93.96, Total loss: 6.822466166139939
Evaluating model epoch_24.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 93.39, Total loss: 7.0754772430412975
Evaluating model epoch_25.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 93.39, Total loss: 7.078118263750239
Evaluating model epoch_26.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 93.65, Total loss: 6.867567944506159
Evaluating model epoch_27.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.08, Total loss: 6.996029606132619
Evaluating model epoch_28.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.46, Total loss: 6.478168967785946
Evaluating model epoch_29.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.17, Total loss: 6.792778887691657
Evaluating model epoch_30.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 93.99, Total loss: 6.771192045031031
Evaluating model epoch_31.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 93.99, Total loss: 6.55683923764395
Evaluating model epoch_32.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.40, Total loss: 7.0021849906406874
Evaluating model epoch_33.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.11, Total loss: 6.823320313338124
Evaluating model epoch_34.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.60, Total loss: 6.845679462062082
Evaluating model epoch_35.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.69, Total loss: 6.445111284011591
Evaluating model epoch_36.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.69, Total loss: 6.5078572167167055
Evaluating model epoch_37.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.40, Total loss: 6.546404130408215
Evaluating model epoch_38.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.22, Total loss: 6.78484349085917
Evaluating model epoch_39.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.37, Total loss: 6.59082422368728
Evaluating model epoch_40.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 93.91, Total loss: 6.69702909516799
Evaluating model epoch_41.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.20, Total loss: 6.76764812216952
Evaluating model epoch_42.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 93.59, Total loss: 6.82570722080038
Evaluating model epoch_43.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.31, Total loss: 6.688829992071319
Evaluating model epoch_44.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.60, Total loss: 6.49662119243688
Evaluating model epoch_45.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.74, Total loss: 6.393234942167835
Evaluating model epoch_46.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.80, Total loss: 6.323944596037969
Evaluating model epoch_47.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.74, Total loss: 6.5487193268044885
Evaluating model epoch_48.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.40, Total loss: 6.582913077187252
Evaluating model epoch_49.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.40, Total loss: 6.362560008054834
Evaluating model epoch_50.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.51, Total loss: 6.287022668641224
Evaluating model epoch_51.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.89, Total loss: 6.161709469521196
Evaluating model epoch_52.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.37, Total loss: 6.779750999248908
Evaluating model epoch_53.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.28, Total loss: 7.682531921938868
Evaluating model epoch_54.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 95.06, Total loss: 6.150015494908922
Evaluating model epoch_55.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.74, Total loss: 6.251866921847764
Evaluating model epoch_56.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.60, Total loss: 6.8822483049616325
Evaluating model epoch_57.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.83, Total loss: 6.273287522037485
Evaluating model epoch_58.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.80, Total loss: 6.297315526411813
Evaluating model epoch_59.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.66, Total loss: 6.6831805428165145
Evaluating model epoch_60.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 95.03, Total loss: 6.206557267115662
Evaluating model epoch_61.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.66, Total loss: 6.893292412047079
Evaluating model epoch_62.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.60, Total loss: 6.611815895889569
Evaluating model epoch_63.pt


Validating:   0%|          | 0/55 [00:00<?, ?it/s]

Test accuracy: 94.92, Total loss: 6.890223489045573
Best accuracy: 95.06208489748774 at epoch_54.pt
Best loss: 6.150015494908922 at epoch_54.pt


# Small

Best accuracy: 90.70170372509385 at epoch_14.pt

Best loss: 10.261200130774341 at epoch_7.pt

# Large

Best accuracy: 95.06208489748774 at epoch_54.pt

Best loss: 6.150015494908922 at epoch_54.pt